app/domain/authentication/authn_k8s/authenticator.rb

```diff
@@ -0,0 +1,300 @@
+        #
+        # return true for valid credentials, false otherwise
+      end
+
+      private
+
+      ### authn-k8s LoginController helpers
+
+      def install_signed_cert cert
+        exec = KubectlExec.new @pod, container: k8s_container_name
+        response = exec.copy "/etc/conjur/ssl/client.pem", cert.to_pem, "0644"
+        
+        if response[:error].present?
+          raise AuthenticationError, response[:error].join
+        end
+      end
+
+      def host_id_param
+        if !@host_id_param
+          cn_entry = get_subject_hash(pod_csr)["CN"]
+          raise CSRVerificationError, 'CSR must contain CN' unless cn_entry
+
+          @host_id_param = cn_entry.gsub('.', '/')
+        end
+
+        @host_id_param
+      end
+
+      def spiffe_id
+        @spiffe_id ||= csr_spiffe_id(pod_csr)
+      end
+
+      def pod_name
+        if !@pod_name
+          raise CSRVerificationError, 'CSR must contain SPIFFE ID SAN' unless spiffe_id
+
+          _, _, namespace, _, @pod_name = URI.parse(spiffe_id).path.split("/")
+          raise CSRVerificationError, 'CSR SPIFFE ID SAN namespace must match conjur host id namespace' unless namespace == k8s_namespace
+        end
+
+        @pod_name
+      end
+
+      def pod_csr
+        if !@pod_csr
+          @pod_csr = OpenSSL::X509::Request.new @request.body.read
+          raise CSRVerificationError, 'CSR can not be verified' unless @pod_csr.verify @pod_csr.public_key
+        end
+
+        @pod_csr
+      end
+
+      # ssl stuff
+
+      def csr_spiffe_id(csr)
+        # https://stackoverflow.com/questions/46494429/how-to-get-an-attribute-from-opensslx509request
+        attributes = csr.attributes
+        raise CSRVerificationError, "CSR must contain workload SPIFFE ID subjectAltName" if not attributes
+
+        seq = nil
+        values = nil
+
+        attributes.each do |a|
+          if a.oid == 'extReq'
+            seq = a.value
+            break
+          end
+        end
+        raise CSRVerificationError, "CSR must contain workload SPIFFE ID subjectAltName" if not seq
+
+        seq.value.each do |v|
+          v.each do |v|
+            if v.value[0].value == 'subjectAltName'
+              values = v.value[1].value
+              break
+            end
+            break if values
+          end
+        end
+        raise CSRVerificationError, "CSR must contain workload SPIFFE ID subjectAltName" if not values
+
+        values = OpenSSL::ASN1.decode(values).value
+
+
+        uris = begin
+                 URI_from_asn1_seq(values)
+               rescue StandardError => e
+                 raise CSRVerificationError, e.message
+               end
+
+        raise CSRVerificationError, "CSR must contain exactly one URI SAN" unless (uris.count == 1)
+        uris[0]
+      end
+
+      ### authn-k8s ApplicationController helpers
+      
+      def verify_enabled
+        conjur_authenticators = (@env['CONJUR_AUTHENTICATORS'] || '').split(',').map(&:strip)
```

Comment:
Consider using `::Authentication::Webservices.from_string` here if possible... this essence of the logic here and there are the same.